composites/Plugin/SnippetPreview/components/SnippetPreview.js

```diff
@@ -1,17 +1,625 @@
+	 *
+	 * Has a leeway of 2 to make sure weird measurements don't cause an infinite
+	 * loop.
+	 *
+	 * @param {HTMLElement} element The element to check.
+	 *
+	 * @returns {boolean} Whether it has content that doesn't fit.
+	 */
+	hasOverflowedContent( element ) {
+		return Math.abs( element.clientHeight - element.scrollHeight ) >= 2;
+	}
+
+	/**
+	 * Set the title in the state so it fits in two lines.
+	 *
+	 * @returns {void}
+	 */
+	fitTitle() {
+		const titleElement = this._titleElement;
+
+		// When the title is 600 pixels it approximately fits 200 characters.
+		const TWO_LINES_OF_CHARACTERS_PER_WIDTH = 3;
+
+		if ( this.hasOverflowedContent( titleElement ) ) {
+			let prevTitle = this.state.title;
+
+			// Heuristic to prevent too many re-renders.
+			const maxCharacterCount = titleElement.clientWidth / TWO_LINES_OF_CHARACTERS_PER_WIDTH;
+
+			if ( prevTitle.length > maxCharacterCount ) {
+				prevTitle = prevTitle.substring( 0, maxCharacterCount );
+			}
+
+			const newTitle = this.dropLastWord( prevTitle );
+
+			this.setState( {
+				title: newTitle,
+			} );
+		}
+	}
+
+	/**
+	 * Set the description in the state so it fits in four lines.
+	 *
+	 * @returns {void}
+	 */
+	fitDescription() {
+		const descriptionElement = this._descriptionElement;
+		const FOUR_LINES_OF_CHARACTERS_PER_WIDTH = 1.75;
+
+		if ( this.hasOverflowedContent( descriptionElement ) ) {
+			let prevDescription = this.state.description;
+
+			// Heuristic to prevent too many re-renders.
+			const maxCharacterCount = descriptionElement.clientWidth / FOUR_LINES_OF_CHARACTERS_PER_WIDTH;
+
+			if ( prevDescription.length > maxCharacterCount ) {
+				prevDescription = prevDescription.substring( 0, maxCharacterCount );
+			}
+
+			const newDescription = this.dropLastWord( prevDescription );
+
+			this.setState( {
+				description: newDescription,
+			} );
+		}
+	}
+
+	/**
+	 * Removes the last word of a sentence.
+	 *
+	 * @param {string} sentence The sentence to drop a word of.
+	 * @returns {string} The new sentence.
+	 */
+	dropLastWord( sentence ) {
+		let titleParts = sentence.split( " " );
+		titleParts.pop();
+
+		return titleParts.join( " " );
+	}
+
+	/**
+	 * Returns the title for rendering.
+	 *
+	 * @returns {string} The title to render.
+	 */
+	getTitle() {
+		if ( this.props.title !== this.state.title ) {
+			return this.state.title + " ...";
+		}
+
+		return this.props.title;
+	}
+
+	/**
+	 * Renders the description for display
+	 *
+	 * @returns {ReactElement} The rendered description.
+	 */
+	renderDescription() {
+		if ( this.props.mode === MOBILE && this.props.description !== this.state.description ) {
+			return this.state.description + " ...";
+		}
+
+		if ( this.props.mode === DESKTOP ) {
+			return truncate( this.props.description, {
+				length: DESCRIPTION_LIMIT,
+				omission: "",
+			} );
+		}
+
+		return this.props.description;
+	}
+
+	/**
+	 * Renders the date if set.
+	 *
+	 * @returns {?ReactElement} The rendered date.
```

Comment:
Is this an intentional pattern? If so, cool!
If not, please change :).

Reply:
Yes, that is intentional. The return value is 'optional'.